crypto/fipsmodule/service_indicator/service_indicator_test.cc

```diff
@@ -0,0 +1,168 @@
+    0x1f, 0xc3, 0x8c, 0xb8, 0x98, 0x86, 0xa3, 0xe3, 0x6c, 0xfc, 0xad,
+    0x3a, 0xb5, 0x59, 0x27, 0x7d, 0x21, 0x07, 0xca, 0x4c, 0x1d, 0x55,
+    0x34, 0xdd, 0x5a, 0x2d, 0xc4, 0xb4, 0xf5, 0xa8,
+#if !defined(BORINGSSL_FIPS_BREAK_AES_CBC)
+    0x35
+#else
+    0x00
+#endif
+};
+
+TEST(ServiceIndicatorTest, BasicTest) {
+  // Reset and check the initial state and counter.
+  awslc_fips_service_indicator_reset_state();
+  struct fips_service_indicator_state *state = awslc_fips_service_indicator_get_state();
+  ASSERT_TRUE(state);
+  ASSERT_EQ(state->counter,0);
+  ASSERT_EQ(state->serviceID, fips_approved_no_state);
+
+  int counter = awslc_fips_service_indicator_get_counter();
+  ASSERT_EQ(counter, state->counter);
+
+  // Call a random approved service.
+  EVP_AEAD_CTX aead_ctx;
+  EVP_AEAD_CTX_zero(&aead_ctx);
+  uint8_t output[256];
+  size_t out_len;
+  EVP_AEAD_CTX_init(&aead_ctx, EVP_aead_aes_128_gcm_randnonce(), kAESKey,
+                         sizeof(kAESKey), 0, nullptr);
+  EVP_AEAD_CTX_seal(&aead_ctx, output, &out_len, sizeof(output), nullptr,
+             0, kPlaintext, sizeof(kPlaintext), nullptr, 0);
+
+  ASSERT_TRUE(awslc_fips_check_service_approved(counter, fips_approved_evp_aes_128_gcm));
+
+  // Check state and counter after using an approved service.
+  state = awslc_fips_service_indicator_get_state();
+  ASSERT_TRUE(state);
+  ASSERT_EQ(state->counter,1);
+  ASSERT_EQ(state->serviceID, fips_approved_evp_aes_128_gcm);
+
+  counter = awslc_fips_service_indicator_get_counter();
+  ASSERT_EQ(counter, state->counter);
+}
+
+TEST(ServiceIndicatorTest, AESCBC) {
+  int counter = awslc_fips_service_indicator_get_counter();
+
+  AES_KEY aes_key;
+  uint8_t aes_iv[16];
+  uint8_t output[256];
+
+  // AES-CBC Encryption KAT
+  memcpy(aes_iv, kAESIV, sizeof(kAESIV));
+  if (AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0) {
+    fprintf(stderr, "AES_set_encrypt_key failed.\n");
+    return;
+  }
+  AES_cbc_encrypt(kPlaintext, output, sizeof(kPlaintext), &aes_key, aes_iv,
+                  AES_ENCRYPT);
+  if (!check_test(kAESCBCCiphertext, output, sizeof(kAESCBCCiphertext),
+                  "AES-CBC Encryption KAT")) {
+    return;
+  }
+
+  ASSERT_TRUE(awslc_fips_check_service_approved(counter, fips_approved_evp_aes_128_cbc));
+  counter = awslc_fips_service_indicator_get_counter();
+
+  // AES-CBC Decryption KAT
+  memcpy(aes_iv, kAESIV, sizeof(kAESIV));
+  if (AES_set_decrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0) {
+    fprintf(stderr, "AES_set_decrypt_key failed.\n");
+    return;
+  }
+  AES_cbc_encrypt(kAESCBCCiphertext, output, sizeof(kAESCBCCiphertext),
+                  &aes_key, aes_iv, AES_DECRYPT);
+  if (!check_test(kPlaintext, output, sizeof(kPlaintext),
+                  "AES-CBC Decryption KAT")) {
+    return;
+  }
+
+  ASSERT_TRUE(awslc_fips_check_service_approved(counter, fips_approved_evp_aes_128_cbc));
+}
+
+TEST(ServiceIndicatorTest, AESGCM) {
+  EVP_AEAD_CTX aead_ctx;
```

Comment:
Should let C++ manage memory e.g. https://github.com/awslabs/aws-lc/blob/main/include/openssl/aead.h#L471